misk/src/main/kotlin/misk/client/ClientInvocationHandler.kt

```diff
@@ -64,11 +72,34 @@ internal class ClientInvocationHandler(
 
     if (tracer != null) retrofitBuilder.callFactory(TracingCallFactory(actionSpecificClient, tracer))
 
+    val mediaTypes = getEndpointMediaTypes(methodName)
+    if (mediaTypes.contains(MediaTypes.APPLICATION_PROTOBUF)) {
+      retrofitBuilder.addConverterFactory(WireConverterFactory.create())
+    }
+    // Always add JSON as default. Ensure it's added last so that other converters get a chance since
+    // JSON converter will accept any object.
+    retrofitBuilder.addConverterFactory(MoshiConverterFactory.create(moshi))
+
     methodName to retrofitBuilder
         .build()
         .create(interfaceType.java)
   }.toMap()
 
+  private fun getEndpointMediaTypes(methodName: String): List<String> {
+    val headers =
+        interfaceType.memberFunctions.find { it.name == methodName }?.findAnnotation<Headers>()
+            ?: return emptyList()
+
+    return headers.value.mapNotNull {
+      val (headerKey, headerValue) = it.split(":").map { it.trim() }
+      if (headerKey.equals("Accept", true) || headerKey.equals("Content-type", true)) {
```

Comment:
what happens if both header values are present and different? does retrofit even allow that?

Reply:
It's atypical but allowed. An endpoint is free to send requests in format A and expect a response in format B. Retrofit supports that by allowing multiple converters, which it iterates through in order looking for one that supports the object it's attempting to convert.